File: api/src/main/java/com/bindglam/neko/api/config/CustomItemPropertiesConfigurable.java
package com.bindglam.neko.api.config;

import com.bindglam.neko.api.content.item.CustomItemProperties;
import io.papermc.paper.registry.RegistryAccess;
import io.papermc.paper.registry.RegistryKey;
import org.bukkit.configuration.ConfigurationSection;
import org.jetbrains.annotations.Nullable;

import java.util.Objects;

public class CustomItemPropertiesConfigurable implements Configurable<CustomItemProperties, ConfigurationSection> {
    @Override
    public @Nullable CustomItemProperties load(@Nullable ConfigurationSection config) {
        if(config == null) return null;

        return new CustomItemProperties(
                Objects.requireNonNull(RegistryAccess.registryAccess().getRegistry(RegistryKey.ITEM).get(Objects.requireNonNull(Configurable.KEY.load(config.getString("type")))), "itemType cannot be null"),
                Configurable.COMPONENT.load(config.getString("display-name")),
                Configurable.COMPONENT_LIST.load(config.getStringList("lore")),
                Configurable.KEY.load(config.getString("item-model"))
        );
    }
}

File: core/src/main/kotlin/com/bindglam/neko/config/CustomItemPropertiesConfigurable.kt
package com.bindglam.neko.config

import com.bindglam.neko.api.config.Configurable
import com.bindglam.neko.api.content.item.CustomItemProperties
import com.bindglam.neko.utils.COMPONENT_CONFIGURABLE
import com.bindglam.neko.utils.ITEM_TYPE_CONFIGURABLE
import com.bindglam.neko.utils.KEY_CONFIGURABLE
import org.bukkit.configuration.ConfigurationSection

class CustomItemPropertiesConfigurable : Configurable<CustomItemProperties, ConfigurationSection> {
    override fun load(config: ConfigurationSection?): CustomItemProperties? = config?.let { CustomItemProperties(
        ITEM_TYPE_CONFIGURABLE.load(config.getString("type"))!!,
        COMPONENT_CONFIGURABLE.load(config.getString("display-name")),
        Configurable.parseAsList(config.getStringList("lore"), COMPONENT_CONFIGURABLE),
        KEY_CONFIGURABLE.load(config.getString("item-model"))
    ) }
}